### PR TITLE
Refactor `fix_trailing_whitespace`

### LINKED ISSUE #223: Add tests for fix-trailing-whitespace

### LINKED ISSUE #294: Avoid reading whole file into memory in `fix_trailing_whitespace` and `fix_end_of_file` hooks

### PR DESCRIPTION
**Pull request by tmdhhl**
Closes #294
Closes #223 

This PR refactors `fix_trailing_whitespace.rs`, improves the handling of trailing whitespace, especially for Markdown files where exactly two trailing spaces at the end of a line indicate a line break. Related to #294. 

## Main changes

- Introduced a `Chars` type to encapsulate a custom list of characters to trim, resolving an issue where `clap` cannot correctly parse `--chars= \t` into `Vec<char>`.
- Optimized the `fix_file` function logic:
  - Only preserve and normalize to exactly two trailing spaces on non-empty Markdown lines that **already have two trailing spaces**; it does not add missing trailing spaces.
  - Normalize lines with more than two trailing spaces to exactly two spaces.
  - Remove all trailing whitespace on blank lines or non-Markdown files, with no extra spaces preserved.
- Added extensive unit tests covering:
  - Markdown lines with excessive or insufficient trailing spaces.
  - Forced Markdown mode.
  - Custom trim character sets.
  - Different line ending formats (including CRLF).
  - Files without trailing newline.
  - Wildcard Markdown extensions.
- Only write back files if actual content changes occur, avoiding unnecessary writes.

## About `Chars` type encapsulation and `clap` parsing issue

When using `clap` to parse command-line arguments, directly parsing a parameter like `--chars=" \t"` into a `Vec<char>` runs into issues because:

- `clap` splits arguments by whitespace by default, causing strings containing spaces and tabs to be misparsed or treated as a single element.
- For example, passing `--chars=" \t"` may not correctly convert into a character vector containing space and tab characters.

To solve this, we implemented a custom `Chars` type with the `FromStr` trait:

- `Chars` wraps a `Vec<char>` internally and manually converts the input string into a vector of characters.
- This approach allows correct parsing regardless of spaces, tabs, or other special characters included in the input.

This workaround bypasses `clap`'s default parsing limitations and enables flexible handling of user-defined trim character sets.

## Trailing two-space preservation rules and reference

- **Only lines that already have exactly two trailing spaces will preserve these spaces as Markdown line breaks.**
- Markdown lines without two trailing spaces will have their trailing whitespace removed normally, without automatically adding spaces.
- This rule prevents unnecessary modifications and aligns with Markdown syntax standards.

This behavior is consistent with the original implementation in the open-source project pre-commit. For example:

```python
@pytest.mark.parametrize(
    ('input_s', 'expected'),
    (
        ('foo \nbar \n', 'foo\nbar\n'),
        ('bar\t\nbaz\t\n', 'bar\nbaz\n'),
    ),
)
def test_fixes_trailing_whitespace(input_s, expected, tmpdir):
    path = tmpdir.join('file.md')
    path.write(input_s)
    assert main((str(path),)) == 1
    assert path.read() == expected
```

### CODE REVIEW

File: src/builtin/pre_commit_hooks/fix_trailing_whitespace.rs [hunk 1]
```diff
@@ -128,3 +97,361 @@ pub(crate) async fn fix_trailing_whitespace(
...
+    let is_markdown = force_markdown || ext.is_some_and(|e| markdown_exts.contains(&e));
+
+    let named_temp_file = NamedTempFile::new()?;
+    let temp_file = tokio::fs::File::from_std(named_temp_file.reopen()?);
+    let mut buf_writer = BufWriter::new(temp_file);
+
+    let file = fs_err::tokio::File::open(filename).await?;
+    let mut buf_reader = BufReader::new(file);
+    let mut line = String::new();
+    let mut modified = false;
```
[the remark this answers is not in the document]
Comment (j178) [reply src/builtin/pre_commit_hooks/fix_trailing_whitespace.rs:120]: I think this is valid, we can avoid creating temp files for every input files.

[the remark this answers is not in the document]
Comment (tmdhhl) [reply src/builtin/pre_commit_hooks/fix_trailing_whitespace.rs:120]: If we don’t want to create a temporary file for each input file, we have to cache the content of the current file, which in the worst case means reading the entire file into memory. Alternatively, we could reuse the same temporary file repeatedly, but that would lose the advantage of concurrent processing.

Which approach do you prefer, or do you think there is a more suitable way to handle this?

[the remark this answers is not in the document]
Comment (j178) [reply src/builtin/pre_commit_hooks/fix_trailing_whitespace.rs:120]: Got it! How about we first check the file size? If it’s less than 16KB, we can read it straight into memory. But if it’s bigger, let’s make a temporary file for it instead. What do you think?

[the remark this answers is not in the document]
Comment (tmdhhl) [reply src/builtin/pre_commit_hooks/fix_trailing_whitespace.rs:120]: Good idea！I will implement it.

[the remark this answers is not in the document]
Comment (tmdhhl) [reply src/builtin/pre_commit_hooks/fix_trailing_whitespace.rs:120]: @j178 The implementation logic is as follows, please review it.

- For small files (≤16KB), use in-memory buffer to avoid IO overhead
- For large files (>16KB), use temporary file buffer to efficiently handle data without excessive memory use
- Implement factory function to select appropriate buffer type dynamically
- Abstract buffer operations behind async trait for consistent interface

[the remark this answers is not in the document]
Comment (j178) [reply src/builtin/pre_commit_hooks/fix_trailing_whitespace.rs:120]: Great, this looks elegant!
